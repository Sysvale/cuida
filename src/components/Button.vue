<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<button
		v-cdstip="tooltipDisabled"
		class="button__container"
		:class="computedStyle"
		@click.stop="clickHandler"
	>
		<cds-spinner
			v-if="loading"
			variant="white"
			size="sm"
			class="button__prepend"
		/>
		<div
			v-if="hasPrependSlot && !loading"
			class="button__prepend"
		>
			<!-- @slot Slot para exibir prepend do botão. -->
			<slot name="prepend" />
		</div>
		<!-- @slot Slot padrão utilizado para exibir texto do botão. -->
		<slot>
			{{ text }}
		</slot>
		<div
			v-if="hasAppendSlot"
			class="button__append"
		>
			<!-- @slot Slot para exibir append do botão. -->
			<slot name="append" />
		</div>
	</button>
</template>

<script>
import CdsSpinner from '../components/Spinner.vue';
import Cdstip from '../utils/directives/cdstip';

export default {

	directives: {
		cdstip: Cdstip,
	},

	components: {
		CdsSpinner,
	},

	props: {
		/**
		 * A variante de cor. São 9 variantes implementadas: 'green', 'teal',
		 * 'blue', 'indigo', 'violet', 'pink', 'red', 'orange' e 'amber'.
		 */
		variant: {
			type: String,
			default: 'green',
		},
		/**
		 * Especifica o tamanho do botão. São 3 tamanhos implementados: 'sm', 'md', 'lg'.
		 */
		size: {
			type: String,
			default: 'md',
		},
		/**
		* Quando true, torna a largura do botão fluida
		*/
		block: {
			type: Boolean,
			default: false,
		},
		/**
		 * Especifica o texto a ser apresentado no corpo do botão.
		 * Este texto será exibido apenas se o slot default não for utilizado.
		 */
		text: {
			type: String,
			default: 'Lorem ipsum',
		},
		/**
		 * Controla a disponibilidade do Botão.
		 */
		disabled: {
			type: Boolean,
			default: false,
		},
		/**
		 * Texto a ser exibido como tooltip com o hover do botão quando a prop disabled estiver ativa.
		 */
		tooltipText: {
			type: String,
			default: null,
		},
		/**
		 * Especifica se a versão do Botão é a secundária.
		 */
		secondary: {
			type: Boolean,
			default: false,
		},
		/**
		 * Especifica se a versão do Botão é a secundária.
		 */
		loading: {
			type: Boolean,
			default: false,
		},
		/**
		* Especifica se o componente deve ser exibido na sua versão ghost.
		*/
		ghost: {
			type: Boolean,
			default: false,
		},
	},

	data() {
		return {
			predefinedColors: [
				'green',
				'teal',
				'turquoise',
				'blue',
				'indigo',
				'violet',
				'pink',
				'red',
				'orange',
				'amber',
				'dark',
			],
			predefinedSizes: [
				'sm',
				'md',
				'lg',
			],
		};
	},

	computed: {
		widthResolver() {
			return this.block ? '100%' : 'max-content';
		},

		tooltipDisabled() {
			return this.disabled && this.tooltipText !== '' ? this.tooltipText : null;
		},

		predefinedColor() {
			if (this.ghost) {
				return 'button--ghost';
			}

			if (this.secondary) {
				return 'button--secondary';
			}

			if (this.predefinedColors.indexOf(this.variant) > -1) {
				return `button--${this.variant}`;
			}

			return 'button--green';
		},

		predefinedSize() {
			if (this.predefinedSizes.indexOf(this.size) > -1) {
				return `button-size--${this.size}`;
			}
			return 'button-size--md';
		},

		computedStyle() {
			const disabled = this.disabled ? '--disabled' : '--active';

			return `${this.predefinedColor}${disabled} ${this.predefinedSize}`;
		},

		hasPrependSlot(){
			return Object.keys(this.$slots).some(slot => slot === 'prepend');
		},

		hasAppendSlot(){
			return Object.keys(this.$slots).some(slot => slot === 'append');
		},
	},

	methods: {
		clickHandler() {
			if (this.disabled) {
				return;
			}
			/**
			* Evento que indica que o Botão foi clicado
			* @event button-click
			* @type {Event}
			*/
			this.$emit('button-click', true);
		},
	},
};
</script>
<style lang="scss" scoped>
@use 'sass:color';
@use '../assets/sass/tokens/index' as tokens;

.button {
	&--secondary {
		&--active {
			background-color: tokens.$n-10;
			color: tokens.$n-700;
			border: 1px solid tokens.$n-50 !important;
			
			&:hover {
				@extend .button--secondary--active;
				background-color: tokens.$n-20;
			}
		}

		&--disabled {
			cursor: default !important;
			background-color: tokens.$n-10;
			color: tokens.$n-300;
			border: 1px solid tokens.$n-30 !important;
		}
	}

	&--ghost {
		&--active {
			background: none;
			cursor: pointer;
		}

		&:hover {
			@extend .button--ghost--active;
			background-color: $n-10;
		}

		&--disabled {
			cursor: default !important;
			background: none;
			color: $n-300;
		}
	}

	&__prepend {
		margin: mr(3);
	}

	&__loader {
		margin: tokens.mr(3);
	}

	&__append {
		margin: ml(3);
	}

	&__container {
		font-family: Satoshi, Inter, Avenir, Helvetica, Arial, sans-serif;
		display: flex;
		align-items: center;
		justify-content: center;
		width: v-bind(widthResolver);
		border: none;
		cursor: pointer;

		&:focus {
			outline: none !important;
		}
	}

	$colors: (
		'--green': (
			'active': tokens.$gp-400,
			'disabled': tokens.$gp-300,
		),
		'--teal': (
			'active': tokens.$ta-400,
			'disabled': tokens.$ta-300,
		),
		'--turquoise': (
			'active': tokens.$ts-400,
			'disabled': tokens.$ts-300,
		),
		'--blue': (
			'active': tokens.$bn-400,
			'disabled': tokens.$bn-300,
		),
		'--indigo': (
			'active': tokens.$in-400,
			'disabled': tokens.$in-300,
		),
		'--violet': (
			'active': tokens.$vr-400,
			'disabled': tokens.$vr-300,
		),
		'--pink': (
			'active': tokens.$pp-400,
			'disabled': tokens.$pp-300,
		),
		'--red': (
			'active': tokens.$rc-400,
			'disabled': tokens.$rc-300,
		),
		'--orange': (
			'active': tokens.$og-400,
			'disabled': tokens.$og-300,
		),
		'--amber': (
			'active': tokens.$al-400,
			'disabled': tokens.$al-300,
		),
		'--dark': (
			'active': tokens.$n-700,
			'disabled': tokens.$n-500,
		),
	);

	@each $color, $variants in $colors {
		&#{$color} {
			@each $state, $disabled-color in $variants {
				&--#{$state} {
					background-color: $disabled-color;
					color: tokens.$n-0;
				}

				@if $state == 'active' {
					&--#{$state}:hover {
						background-color: color.adjust($disabled-color, $lightness: -5%);
						color: tokens.$n-0;
					}
				}
			}

			&--disabled {
				cursor: default;
			}
		}
	}
}

.button-size {
	&--sm {
		@include tokens.button-3;
		padding: tokens.pYX(2, 4);
		height: 32px;
		border-radius: tokens.$border-radius-lil;
	}

	&--md {
		@include tokens.button-2;
		padding: tokens.pYX(3, 6);
		height: 40px;
		border-radius: tokens.$border-radius-extra-small;
	}

	&--lg {
		@include tokens.button-1;
		padding: tokens.pYX(4, 8);
		height: 48px;
		border-radius: tokens.$border-radius-extra-small;
	}
}
</style>
