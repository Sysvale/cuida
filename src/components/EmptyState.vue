<template>
	<div
		class="empty-state"
	>
		<!-- @slot Slot usado para inserção de conteúdo gráfico
			customizado como alternativa a imagem do empty-state. -->
		<slot name="graphic-element">
			<img
				class="empty-state__image"
				:src="image"
				:alt="imageDescription"
			>
		</slot>
		<div class="empty-state__title">
			{{ title }}
		</div>

		<div class="empty-state__text">
			<!-- @slot Slot usado para inserção de conteúdo customizado no texto do corpo do
				empty state. -->
			<slot name="text">
				{{ text }}
			</slot>
		</div>

		<!--
			Evento emitido quando o botão do Empty State é clicado
			@event action-button-click
			@type {Event}
		-->
		<CdsButton
			v-if="!hideActionButton"
			class="empty-state__button"
			:text="actionButtonText"
			:variant="actionButtonVariant"
			@click="$emit('action-button-click', true)"
		/>
	</div>
</template>

<script>
import CdsButton from './Button.vue';

const predefinedColors = [
	'teal',
	'green',
	'blue',
	'indigo',
	'violet',
	'pink',
	'red',
	'orange',
	'amber',
];

export default {
	name: 'CdsEmptyState',
	components: {
		CdsButton,
	},

	props: {
		/**
		 * A imagem a ser mostrada no Empty State. Obs.: o addon Controls
		 * do Storybook ainda não permite a seleção de arquivos. Desse modo
		 * não é possível testar o funcionamento dessa prop por aqui.
		 */
		image: {
			type: String,
			default: '',
		},
		/**
		 * O descritor da imagem do Empty State adicionado à 'alt' da tag <img>
		 */
		imageDescription: {
			type: String,
			default: 'Imagem de Empty State',
		},
		/**
		 * O título do Empty State.
		 */
		title: {
			type: String,
			default: 'Título do empty state',
			required: true,
		},
		/**
		 * O texto instrutivo do Empty State
		 */
		text: {
			type: String,
			default: 'Para sair dessa situação de empty state, realize a ação abaixo.',
		},
		/**
		 * Prop utilizada para ocultar o botão de ação do Empty State.
		 */
		hideActionButton: {
			type: Boolean,
			default: false,
		},
		/**
		 * O texto mostrado no botão de ação do Empty State
		 */
		actionButtonText: {
			type: String,
			default: 'Finalizar',
		},
		/**
		 * A variante do botão de ação do Empty State (segue as variantes do componente de botão do Cuida)
		 * @values green, teal, blue, indigo, violet, pink, red, orange, amber, dark
		 */
		actionButtonVariant: {
			type: String,
			default: 'green',
			validator: (value) => predefinedColors.includes(value),
		},
	},
};
</script>

<style lang="scss" scoped>
@use '../assets/sass/tokens/index' as tokens;

.empty-state {
	align-items: center;
	display: flex;
	flex-direction: column;
	margin: tokens.mYX(12, 5);
	position: relative;

	&__image {
		height: auto;
		max-height: 180px;
		max-width: 180px;
		width: auto;
	}

	&__title {
		color: tokens.$n-600;
		margin: tokens.mt(8);
		@include tokens.subheading-2;
	}

	&__text {
		color: tokens.$n-500;
		margin: tokens.mt(1);
		max-width: 50%;
		text-align: center;
		@include tokens.body-1;
	}

	&__button {
		margin: tokens.mt(4);
	}
}

@media (max-width: 932px) {
	.empty-state__text {
		max-width: 100%;
	}
}
</style>
