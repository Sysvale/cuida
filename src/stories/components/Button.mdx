import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import Button from '../../components/Button.vue';
import * as ButtonStories from './Button.stories';

<Meta of={ButtonStories} />

export const Template = (args) => ({
	components: { CdsButton: Button },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-button
			v-bind="args"
            @button-click="logButtonClick"
		>
		</cds-button>`,
    methods: {
		logButtonClick(event) {
			console.info('%con click: ', 'color: #2C70CD;', event);
		},
	},
});

# Button

### Buttons são componentes que permitem que o usuário execute uma ação com um toque.

***

<br />

## Quando usar:

* For necessário comunicar ao usuário que ele pode executar uma ação na interface,
  seja em diálogos, janelas modais, formulários, cards, etc.

<br />

## Quando não usar:

* For necessário dar a opção de executar uma ação como um detalhe pequeno na interface.
  Nesse caso, recomendamos o uso de links.

<br />

## Preview

<Canvas>
  <Story of={ButtonStories.Button} />
</Canvas>

<Controls />
