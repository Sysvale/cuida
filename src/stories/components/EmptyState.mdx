import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import EmptyState from '../../components/EmptyState.vue';
import * as EmptyStateStories from './EmptyState.stories';

<Meta of={EmptyStateStories} />

export const Template = (args) => ({
	components: { CdsEmptyState: EmptyState },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-empty-state
			v-bind="args" 
			@actionButtonClick="logButtonClick"
		/>
	`,
	methods: {
		logButtonClick(event) {
			console.info('logClick: ', event);
		}
	},
});

# EmptyState

### EmptyStates são componentes utilizados em momentos instrutivos aos usuários.

***

<br />

## Quando usar:

* For necessário instruir o usuário sobre o funcionamento de uma determinada página.
* Uma página ou componente não possuir conteúdo padrão a ser mostrado.
* Uma determinada ação resultar em erro.

<br />

## Quando não usar:

* Utilizado como um loader.
* Utilizado apenas com título.

<br />

## Preview

<Canvas>
  <Story of={EmptyStateStories.EmptyState} />
</Canvas>

<Controls />
