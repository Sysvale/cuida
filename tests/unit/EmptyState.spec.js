import { mount, createLocalVue } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';
import EmptyState from '../../src/components/EmptyState.vue';
import Button from '../../src/components/Button.vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

test('Component is mounted properly', () => {
	const wrapper = mount(EmptyState, {
		localVue,
		propsData: {
			title: 'EmptyState title',
			text: 'EmptyState text',
		},
	});

	expect(wrapper).toMatchSnapshot();
});

test('if the event is emitted correctly when the button is clicked', async () => {
	const wrapper = mount(EmptyState, {
		localVue,
		propsData: {
			title: 'EmptyState title',
			text: 'EmptyState text',
		},
	});

	await wrapper.findComponent(Button).vm.$emit('click');

	expect(wrapper.emitted()['action-button-click']).toBeTruthy();
	expect(wrapper.emitted()['action-button-click']).toEqual([[true]]);
});

test('if the button is shown', () => {
	const wrapper = mount(EmptyState, {
		localVue,
		propsData: {
			title: 'EmptyState title',
			text: 'EmptyState text',
		},
	});

	expect(wrapper.findAll('.empty-state__button').length).toBe(1);
});

test('if the button is not shown', () => {
	const wrapper = mount(EmptyState, {
		localVue,
		propsData: {
			hideActionButton: true,
			title: 'EmptyState title',
			text: 'EmptyState text',
		},
	});

	expect(wrapper.findAll('.empty-state__button').length).toBe(0);
});
