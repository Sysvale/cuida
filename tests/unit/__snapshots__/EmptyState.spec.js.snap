// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<div class="empty-state"><img src="" alt="Imagem de Empty State" class="empty-state__image">
  <div class="empty-state__title">
    EmptyState title
  </div>
  <div class="empty-state__text">
    EmptyState text
  </div> <span id="cds-button" class="empty-state__button"><button class="button__container button--green--active button-size--md">
		Finalizar
	</button></span>
</div>
`;
